src/lib/packet-create.c

```diff
@@ -458,34 +454,8 @@ write_seckey_body(const pgp_seckey_t *key, const char *passphrase, pgp_output_t
     }
     pgp_push_enc_crypt(output, crypted);
 
-    switch (key->pubkey.alg) {
-    case PGP_PKA_RSA:
-    case PGP_PKA_RSA_ENCRYPT_ONLY:
-    case PGP_PKA_RSA_SIGN_ONLY:
-        if (!pgp_write_mpi(output, key->key.rsa.d) || !pgp_write_mpi(output, key->key.rsa.p) ||
-            !pgp_write_mpi(output, key->key.rsa.q) || !pgp_write_mpi(output, key->key.rsa.u)) {
-            if (rnp_get_debug(__FILE__)) {
-                RNP_LOG("4 x mpi not written - problem");
-            }
-            return false;
-        }
-        break;
-    case PGP_PKA_DSA:
-        if (!pgp_write_mpi(output, key->key.dsa.x))
-            return false;
-        break;
-    case PGP_PKA_ECDSA:
-    case PGP_PKA_EDDSA:
-    case PGP_PKA_SM2:
-    case PGP_PKA_ECDH:
-        if (!pgp_write_mpi(output, key->key.ecc.x))
-            return false;
-        break;
-    case PGP_PKA_ELGAMAL:
-        if (!pgp_write_mpi(output, key->key.elgamal.x))
-            return false;
-        break;
-    default:
+    if (!pgp_write_secret_mpis(output, key)) {
+        RNP_LOG("failed to write MPIs");
         return false;
```

Comment:
Also, I'm aware this is leaky, I've done a lot more work on this function locally that cleans things up a bit.